src/lsli/main.py

```diff
@@ -0,0 +1,449 @@
+        non_na_systems["Time"] = pd.to_datetime(non_na_systems["Time"], format="mixed")
+        non_na_systems.rename(columns={"PWS ID": "PWSID", "Time": "submitted_time"}, inplace=True)
+        non_na_systems["area_type"] = "Approved System"
+
+        #: Only use the most recent approval for each system
+        self.cleaned_systems_dataframe = non_na_systems.sort_values("submitted_time").drop_duplicates(
+            subset="PWSID", keep="last"
+        )
+
+    def load_system_links_from_gsheet(self) -> None:
+        """Load the interactive maps sheet from Google Sheets using a new extractor"""
+
+        module_logger.debug("Loading interactive map links sheet from Google Sheets...")
+        gsheet_extractor = extract.GSheetLoader(self._credentials)
+        self.links = gsheet_extractor.load_specific_worksheet_into_dataframe(
+            self._links_sheet_id, self._links_sheet_name, by_title=True
+        )
+
+    def clean_system_links(self) -> None:
+        """Format the PWSID, rename columns, and log & drop duplicate PWSIDs"""
+
+        module_logger.debug("Cleaning interactive map links data...")
+
+        #: Drop empty rows
+        self.links.replace("", np.nan, inplace=True)
+        self.links.dropna(subset=["PWSID"], inplace=True)
+
+        #: Clean PWSID, drop duplicates, rename columns
+        self.links["PWSID"] = self.links["PWSID"].str.lower().str.strip("utah").astype(int)
+        self.links.rename(columns={"Water Systme Name": "System Name"}, inplace=True)
+        duplicated_links = self.links[self.links["PWSID"].duplicated(keep=False)]
+
+        if not duplicated_links.empty:
+            module_logger.warning(
+                "Duplicate PWSIDs found in the interactive maps sheet: %s",
+                ", ".join(duplicated_links["PWSID"].astype(str).tolist()),
+            )
+            self.duplicate_link_pwsids = {row["System Name"]: row["PWSID"] for _, row in duplicated_links.iterrows()}
+
+        self.links.drop_duplicates(subset="PWSID", keep="last", inplace=True)
+        self.links["area_type"] = "Link"
+        self.links.rename(columns={"Interactive map link": "link"}, inplace=True)
+        self.links = self.links.reindex(columns=["PWSID", "System Name", "link", "area_type"])
+
+    def load_system_geometries(self, service_areas_service_url: str) -> None:
+        """Load the system area geometries from the specified Feature Service URL, converting PWSIDs to ints
+
+        Args:
+            service_areas_service_url (str): Full REST endpoint URL, including the layer number
+        """
+
+        module_logger.debug("Loading system area geometries from %s...", service_areas_service_url)
+        water_service_areas = arcgis.features.FeatureLayer(service_areas_service_url).query(as_df=True)
+        self.cleaned_water_service_areas = water_service_areas[water_service_areas["DWSYSNUM"] != " "].copy()
```

Comment:
this seems a little fragile. Would it be safer to trim the column and check for an empty string? I assume this one space convention is working now but I wonder how consistent it will stay.